### PR TITLE
Feature/parameter references

### PR DESCRIPTION
**Pull request by ralphdoe**
New validations added to references of parameters.

### Description

Added a new validation when references of parameters point to headers instead of parameters components.

### Motivation and Context

There's no ticket about this yet, this is specified in SWOS-72.

### How Has This Been Tested?

There's no ticket about this yet, this is specified in SWOS-72.

## Checklist

Yes, added unit tests for this case.

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [X] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [X] are not breaking changes.

### Documentation
- [ ] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [X] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [ ] My changes can and should be tested by unit and/or integration tests.
- [X] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [X] All new and existing tests passed.

### CODE REVIEW

File: src/plugins/validate-semantic/validators/oas3/refs.js [hunk 1]
```diff
@@ -29,3 +29,31 @@ export const validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions = ()
...
+      return nodes.reduce((acc, node) => {
+        const value = node.node
+        const ref = value.$ref
+
+        if(!ref) {
+          return acc
+        }
+        
+        // Ignore external references
+        if (ref.startsWith("#/")) {
```
Comment (hkosova) [inline src/plugins/validate-semantic/validators/oas3/refs.js:46]: This check is not needed because it's covered by the next `if`.

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 2]
```diff
@@ -206,4 +206,123 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+    it("should return an error when a header incorrectly references a local parameter schema", () => {
+      const spec = {
+        openapi: "3.0.0",
+        paths: {
+          "/foo": {
+            parameters: {
+              myParam: {
+                $ref: "#/components/headers/foo"
+              }
+            },
```
Comment (hkosova) [inline test/unit/plugins/validate-semantic/oas3/refs.js:220]: Should be:
```suggestion
            parameters: [
              {
                $ref: "#/components/headers/foo"
              }
            ],
```

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 3]
```diff
@@ -206,4 +206,123 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+            get: {
+              parameters: {
+                myParam: {
+                  $ref: "#/components/headers/foo"
+                }
+              },
```
Comment (hkosova) [inline test/unit/plugins/validate-semantic/oas3/refs.js:226]: Should be:
```suggestion
              parameters: [
                {
                  $ref: "#/components/headers/foo"
                }
              ],
```

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 4]
```diff
@@ -206,4 +206,123 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+    })
+    
+    it("should return no errors when a header correctly references a local component header", () => {
+      const spec = {
+        openapi: "3.0.0",
+        paths: {
+          "/foo": {
+            parameters: {
+              $ref: "#/components/parameters/foo"
+            },
```
Comment (hkosova) [inline test/unit/plugins/validate-semantic/oas3/refs.js:267]: Should be:
```suggestion
            parameters: [
              {
                $ref: "#/components/parameters/foo"
              }
            ],
```

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 5]
```diff
@@ -206,4 +206,123 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+            get: {
+              parameters: {
+                $ref: "#/components/parameters/foo"
+              },
```
Comment (hkosova) [inline test/unit/plugins/validate-semantic/oas3/refs.js:271]: Should be:
```suggestion
              parameters: [
                {
                  $ref: "#/components/parameters/foo"
                }
              ],
```

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 6]
```diff
@@ -206,4 +206,123 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+    it("should return no errors when a header correctly references a remote component header", () => {
+      const spec = {
+        openapi: "3.0.0",
+        paths: {
+          "/foo": {
+            parameters: {
+              myParam:{
+                $ref: "http://www.google.com/#/components/parameters/foo"
+              }
+            },
```
Comment (hkosova) [inline test/unit/plugins/validate-semantic/oas3/refs.js:301]: Should be:
```suggestion
            parameters: [
              {
                $ref: "http://www.google.com/#/components/parameters/foo"
              }
            ],
```

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 7]
```diff
@@ -206,4 +206,123 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+            get: {
+              parameters: {
+                myParam:{
+                  $ref: "http://www.google.com/#/components/parameters/foo"
+                }
+              },
```
Comment (hkosova) [inline test/unit/plugins/validate-semantic/oas3/refs.js:307]: Should be:
```suggestion
              parameters: [
                {
                  $ref: "http://www.google.com/#/components/parameters/foo"
                }
              ],
```

File: src/plugins/validate-semantic/validators/oas3/refs.js [hunk 8]
```diff
@@ -29,3 +29,31 @@ export const validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions = ()
       }, [])
     })
 }
+
+export const validateOAS3RefsForHeadersReferenceHeadersPositions = () => sys => {
```
Comment (hkosova) [inline src/plugins/validate-semantic/validators/oas3/refs.js:33]: This name does not match the validation performed - we're verifying all parameters not just headers. Please change to

```
validateOAS3ParameterRefsReferenceParameterPositions
```
or similar.

File: src/plugins/validate-semantic/validators/oas3/refs.js [hunk 9]
```diff
@@ -29,3 +29,31 @@ export const validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions = ()
...
+        if(!ref) {
+          return acc
+        }
+        
+        // Ignore external references
+        if (ref.startsWith("#/")) {
+          if (ref.startsWith("#/components/headers")) {
+            acc.push({
+              level: "error",
+              message: `OAS3 response parameter $refs should point to #/components/parameters/... and not #/components/headers/...`,
```
Comment (hkosova) [inline src/plugins/validate-semantic/validators/oas3/refs.js:50]: The word "response" should be removed.

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 10]
```diff
@@ -206,4 +206,123 @@ describe("validation plugin - semantic - oas3 refs", () => {
       return expectNoErrorsOrWarnings(spec)
     })
   })
-})
+
+  describe("$refs for headers must reference a header by position", () => {
+    it("should return an error when a header incorrectly references a local parameter schema", () => {
```
Comment (hkosova) [inline test/unit/plugins/validate-semantic/oas3/refs.js:211]: Test names don't match the validation performed. We're validating all parameters not just header parameters. Please change to
```js
  describe("parameter $refs should not point to header components", () => {
    it("should return an error when a parameter incorrectly references a response header component", () => {
```
or something like that.

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 11]
```diff
@@ -206,4 +206,123 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+          const secondError = allErrors[4]
+          expect(secondError.message).toEqual(`OAS3 response parameter $refs should point to #/components/parameters/... and not #/components/headers/...`)
+          expect(secondError.path).toEqual(["paths","/foo","get","parameters", "myParam", "$ref"])
+          const thirdError = allErrors[5]
+          expect(thirdError.message).toEqual(`OAS3 response parameter $refs should point to #/components/parameters/... and not #/components/headers/...`)
+          expect(thirdError.path).toEqual(["components","parameters", "myParam", "$ref"])
+        })
+    })
+    
+    it("should return no errors when a header correctly references a local component header", () => {
```
Comment (hkosova) [inline test/unit/plugins/validate-semantic/oas3/refs.js:260]:
```suggestion
    it("should return no errors when a parameter correctly references a parameter component", () => {
```

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 12]
```diff
@@ -206,4 +206,123 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+              $ref: "#/components/parameters/foo"
+            }
+          }
+        }
+      }
+
+      return expectNoErrorsOrWarnings(spec)
+    })
+    
+    it("should return no errors when a header correctly references a remote component header", () => {
```
Comment (hkosova) [inline test/unit/plugins/validate-semantic/oas3/refs.js:292]:
```suggestion
    it("should return no errors for external parameter $refs", () => {
```

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 1 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/oas3/refs.js:46]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 2 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js:220]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 3 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js:226]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 4 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js:267]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 5 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js:271]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 6 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js:301]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 7 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js:307]: Done!

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 8 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/oas3/refs.js:33]: Done!

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 9 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/oas3/refs.js:50]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 10 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js:211]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 11 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js:260]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 12 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js:292]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 13]
```diff
@@ -206,4 +206,127 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+              }
+            }
+          }
+        },
+        components: {
+          parameters: {
+            myParam: {
+              $ref: "#/components/headers/foo"
+            }
+          }
```
Comment (hkosova) [inline test/unit/plugins/validate-semantic/oas3/refs.js:240]: Suggestion: add a dummy header component named `foo` to avoid extra `$ref` resolution errors in this test
```suggestion
          },
          headers: {
            foo: {}
          }
```

Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js:243]: Updated! This improved the test case.

**Comment by hkosova** [comment]
please build
